### ISSUE TITLE
[$40] Do not display error until the user has completed the URL address when adding a link

### ISSUE THREAD

**Issue by vic-topcoder** [appirio-tech/connect-app#2038]
### Expected behavior
I would like to see an error only after I completed entering my  URL address in add link section. While I'm typing, I don't want to see an error. Once I'm done typing, if the link is valid, the button should become blue. If I lose focus and there is error, the error message should appear. Let's be smart about the user experience

### Actual behavior
Error is displayed as soon as I start typing in the field. That's bad UX
![screen shot 2018-06-13 at 9 57 25 am](https://user-images.githubusercontent.com/8484061/41366298-7dd23752-6ef0-11e8-8b9b-1acb486194b6.png)

**Comment by vikasrohit** [comment]
@vic-topcoder  I don't think it is straightforward to fix this because we want to immediately show the error when user updates the already entered valid url in a way that makes the URL invalid. If we are okay with showing error message only when Add Link button is clicked, we can achieve what you like here.

**Comment by vikasrohit** [comment]
Moving the issue to community fix milestone because this is not related to the project plan changes.

**Comment by vikasrohit** [comment]
@rmjmccormick Can you please validate the behaviour once? I think it makes the logic complicated if we want not to show the error message when user is entering the value first time and show the error message when user is updating existing value. Please see my [comment](https://github.com/appirio-tech/connect-app/issues/2038#issuecomment-397181500) above.

**Comment by rmjmccormick** [comment]
@vikasrohit I'm guided by @vic-topcoder here as UX specialist, so I'm minded to follow his lead here. However, I agree we shouldn't complicate the logic needlessly. How about displaying the error message upon validation when an obvious keystroke error is made (e.g. ww. or wwww. at the start of the URL or lack of .xxx, .xx or .xx.xx at the end of the URL)  -how straightforward would that be? 

If that is too complex, I am minded to keep as is as I also note  that the valid URL warning disappears after 6 characters at the moment, right?

![image](https://user-images.githubusercontent.com/39598852/42219829-7d3032bc-7ecd-11e8-9311-e67242c2095c.png)

![image](https://user-images.githubusercontent.com/39598852/42219865-91394d84-7ecd-11e8-88eb-0e2441214944.png)

**Comment by vic-topcoder** [comment]
This behavior of triggering a validation error while the user is still typing is called "premature inline validation"

>Another surprising matter is that the potential implementation pitfalls of inline validation haven’t been highlighted much. During our end-user testing we consistently observe two pitfalls that can end up making live inline validation more frustrating than helpful to the user:
>
>1. Premature Inline Validation – the user’s input is marked as invalid before they have finished typing, essentially reprimanding the user for mistakes they haven’t made and prompting them to look for errors that aren’t there.
>2.Outdated Error Messages – error messages aren’t updated and removed “live”, as the user edits their input, resulting in outdated error messages being displayed even though the user has actually fixed the problem.
from https://baymard.com/blog/inline-form-validation and around the web.

![screen shot 2018-07-03 at 9 57 50 am](https://user-images.githubusercontent.com/8484061/42234016-f0769d22-7ea7-11e8-9557-e51855ddfb25.png)

and so on.

**Comment by vikasrohit** [comment]
Being developer, seems like I am able to understand the criticality of the issue. So, @gondzo please go ahead to get it implemented as per requirements.

**Comment by simranb86** [comment]
I cannot assign this to me - can someone assign it to me ? I will start work on this

**Comment by simranb86** [comment]
@zjuasmn I have been getting 404, tried 3 times setting up whole project but same error always when following README file. Can you confirm README is up-to-date and not missing something?

**Comment by amitpatra** [comment]
I am picking up this issue, if possible please assign me. I don't have repo access. Thanks.

**Comment by andela-jomadoye** [comment]
@sandipanghos  You should not pickup issue and work on them when some with no access to repo(@amitpatra ) has already commented in issue about assigning the issue to them. I believe that it is against the rule as specified in the thread(https://apps.topcoder.com/forums/?module=Thread&threadID=920569&start=0). Thanks.

**Comment by rmjmccormick** [comment]
@gondzo quick question - has this been tested? I have entered a manifestly incorrect URL and it has not triggered any error message at all? What steps need to be reproduced to trigger an error?

![image](https://user-images.githubusercontent.com/39598852/42500139-c91120a2-8430-11e8-811e-b6ed162d68dd.png)

![image](https://user-images.githubusercontent.com/39598852/42500309-313cda4a-8431-11e8-8e53-cc378d4e3730.png)

**Comment by gondzo** [comment]
@rmjmccormick we just validate if the form or url is correct not if it exists or is reachable. It's done by a regex only - https://github.com/appirio-tech/react-components/blob/252c8b5552e52e4116a1cc2b6ef2a508f17a752c/components/Formsy/index.js#L14

To get the error message when adding a link, enter `ttt` into the address field and focus away from the field
![image](https://user-images.githubusercontent.com/3051646/42501066-291939a6-8433-11e8-9cfd-7952a1330f12.png)

**Comment by rmjmccormick** [comment]
@gondzo Understood, However, why is no error flagged when I input ww.marvelapp.comm as in this example both form and URL are incorrect - i.e. ww instead of www and .comm instead of .com. 
The likelihood of someone typing in ttt is very remote; on the other hand, the likelihood of mis-keying the www. prefix or the relevant gTLD is high

**Comment by gondzo** [comment]
`ww.marvelapp.comm` has the correct form (ww could be a subdomain of marvelapp.comm), www prefix isn't required in the urls, comm could be a real comain so we can't throw validation error.
We could implement hints (`did you mean to type xx.yy`), but we can't throw validation errors for entering valid urls. Not sure if we should implement that though, it's nice to have it, but it's not very important

**Comment by vikasrohit** [comment]
Agreed with @gondzo. @rmjmccormick we relaxed the URL validation long ago https://github.com/appirio-tech/connect-app/issues/626 because users are using it store different types of links which are not typically following the common URL patterns.
